catroid/src/main/java/org/catrobat/catroid/web/WebConnection.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.web;
+
+import com.google.common.collect.ImmutableList;
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.Callback;
+import com.squareup.okhttp.ConnectionSpec;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.net.MalformedURLException;
+import java.net.URL;
+
+public class WebConnection {
+	private final OkHttpClient okHttpClient;
+	private static final String USER_AGENT = "Mozilla/5.0 (compatible; Catrobatbot/1.0; +https://catrob.at/bot)";
+
+	private Call call;
+	private String url;
+	private WeakReference<WebRequestListener> weakListenerReference;
+
+	public WebConnection() {
+		okHttpClient = new OkHttpClient();
+		ConnectionSpec modernTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS).build();
+		ConnectionSpec compatibleTlsSpec = new ConnectionSpec.Builder(ConnectionSpec.COMPATIBLE_TLS).build();
+		ConnectionSpec cleartextSpec = new ConnectionSpec.Builder(ConnectionSpec.CLEARTEXT).build();
+		okHttpClient.setConnectionSpecs(ImmutableList.of(modernTlsSpec, compatibleTlsSpec, cleartextSpec));
+	}
+
+	public WebConnection(OkHttpClient okHttpClient) {
+		this.okHttpClient = okHttpClient;
+	}
+
+	public void setListener(WebRequestListener listener) {
+		this.weakListenerReference = new WeakReference<>(listener);
+	}
+
+	public void setUrl(String url) {
+		this.url = url;
+	}
+
+	private synchronized WebRequestListener getListenerAndSetReferenceToNull() {
+		if (this.weakListenerReference != null) {
+			WebRequestListener listener = this.weakListenerReference.get();
+			this.weakListenerReference = null;
+			return listener;
+		}
+		return null;
+	}
+
+	public synchronized boolean sendWebRequest() {
+		try {
+			checkIfUrlIsValid();
+		} catch (RuntimeException e) {
+			return false;
+		}
+
+		Request request = new Request.Builder()
+				.url(url)
+				.header("User-Agent", USER_AGENT)
+				.build();
+
+		call = okHttpClient.newCall(request);
+		call.enqueue(getCallback());
+		return true;
+	}
+
+	private Callback getCallback() {
+		return new Callback() {
+			@Override
+			public void onFailure(Request request, IOException e) {
+			}
+
+			@Override
+			public void onResponse(Response response) throws IOException {
+				WebRequestListener listener = getListenerAndSetReferenceToNull();
+				if (listener != null) {
+					if (response.isSuccessful()) {
+						listener.onRequestFinished(response.body().string());
+					} else {
+						listener.onRequestFinished(String.valueOf(response.code()));
+					}
+				}
+			}
+		};
+	}
+
+	public void cancelCall() {
+		if (call != null) {
+			okHttpClient.getDispatcher().getExecutorService().execute(new Runnable() {
```

Comment:
why does this have to be run as a thread on some okHttpClient provided stack?

Reply:
Calling call.cancel() directly may result in an:
_uncaughtException: 
    android.os.NetworkOnMainThreadException_
Network operations are not allowed to run in the main thread, thus the Runnable in cancelCall().